field/constructed_tlv.go

```diff
@@ -0,0 +1,524 @@
+		f.subfields = CreateSubfields(f.spec)
+	}
+	f.setSubfields = make(map[string]struct{})
+}
+
+// Spec returns the receiver's spec.
+func (f *ConstructedTLV) Spec() *Spec {
+	return f.spec
+}
+
+// describe returns tlv tree with tag and value
+func (f *ConstructedTLV) describe(output io.Writer, padLeft int) {
+
+	if output == nil {
+		output = os.Stdout
+	}
+
+	for _, tag := range f.orderedSpecFieldTags {
+		field, ok := f.subfields[tag]
+		if !ok {
+			continue
+		}
+
+		if field.Spec() == nil || field.Spec().Tag == nil {
+			continue
+		}
+
+		fmtStr := "%s\t:"
+		for i := 0; i < padLeft; i++ {
+			// spaces for tree levels
+			fmtStr = "  " + fmtStr
+		}
+
+		if sub, ok := field.(*PrimitiveTLV); ok {
+			raw, _ := sub.Bytes()
+			fmt.Fprintf(output, fmtStr+" %X\n", field.Spec().Tag.Tag, raw)
+		} else if sub, ok := field.(*ConstructedTLV); ok {
+			fmt.Fprintf(output, fmtStr+"\n", field.Spec().Tag.Tag)
+			sub.describe(output, padLeft+1)
+		}
+	}
+}
+
+// getSubfields returns the map of set sub fields
+func (f *ConstructedTLV) getSubfields() map[string]Field {
+	fields := map[string]Field{}
+	for i := range f.setSubfields {
+		fields[i] = f.subfields[i]
+	}
+	return fields
+}
+
+// SetSpec validates the spec and creates new instances of Subfields defined
+// in the specification.
+// NOTE: Composite does not support padding on the base spec. Therefore, users
+// should only pass None or nil values for ths type. Passing any other value
+// will result in a panic.
+func (f *ConstructedTLV) SetSpec(spec *Spec) {
+	if err := validateConstructedTLVSpec(spec); err != nil {
+		panic(err)
+	}
+	f.spec = spec
+	f.orderedSpecFieldTags = orderedKeys(spec.Subfields, spec.Tag.Sort)
+}
+
+func (f *ConstructedTLV) Unmarshal(v interface{}) error {
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return errors.New("data is not a pointer or nil")
+	}
+
+	// get the struct from the pointer
+	dataStruct := rv.Elem()
+
+	if dataStruct.Kind() != reflect.Struct {
+		return errors.New("data is not a struct")
+	}
+
+	// iterate over struct fields
+	for i := 0; i < dataStruct.NumField(); i++ {
+		indexOrTag, err := getFieldIndexOrTag(dataStruct.Type().Field(i))
+		if err != nil {
+			return fmt.Errorf("getting field %d index: %w", i, err)
+		}
+
+		// skip field without index
+		if indexOrTag == "" {
+			continue
+		}
+
+		messageField, ok := f.subfields[indexOrTag]
+		if !ok {
+			continue
+		}
+
+		// unmarshal only subfield that has the value set
+		if _, set := f.setSubfields[indexOrTag]; !set {
+			continue
+		}
+
+		dataField := dataStruct.Field(i)
+		if dataField.IsNil() {
+			dataField.Set(reflect.New(dataField.Type().Elem()))
+		}
+
+		err = messageField.Unmarshal(dataField.Interface())
+		if err != nil {
+			return fmt.Errorf("failed to get data from field %s: %w", indexOrTag, err)
+		}
+	}
+
+	return nil
+}
+
+// Deprecated. Use Marshal instead
+func (f *ConstructedTLV) SetData(v interface{}) error {
+	return f.Marshal(v)
+}
```

Comment:
Can we remove these deprecated fields? Or do we need them to meet an interface?

Reply:
SetData is still in field interface with deprecated description 
we need to remove this function when getting major change